### PR TITLE
HDDS-3137. OM RpcClient fail with java.lang.IllegalArgumentException.

### PR DESCRIPTION
**Pull request by bharatviswa504**
## What changes were proposed in this pull request?

Fix Rpcclient creation failing with java.lang.IllegalArgumentException

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3137

## How was this patch tested?

Previously in docker-compose om-ha-s3, when one of the OM is killed using docker stop <<containerid>> used to fail with IllegalArgumentException.

Now with the fix, tried on the docker-compose ha cluster.
```
$ docker ps
CONTAINER ID        IMAGE                            COMMAND                  CREATED             STATUS              PORTS                                            NAMES
4181d2fb54bb        apache/ozone-runner:20191107-1   "/usr/local/bin/dumb…"   31 minutes ago      Up 31 minutes       0.0.0.0:9878->9878/tcp                           ozone-om-ha-s3_s3g_1
81e613bda640        apache/ozone-runner:20191107-1   "/usr/local/bin/dumb…"   31 minutes ago      Up 31 minutes       0.0.0.0:9890->9872/tcp, 0.0.0.0:9880->9874/tcp   ozone-om-ha-s3_om1_1
cdd968cba61f        apache/ozone-runner:20191107-1   "/usr/local/bin/dumb…"   31 minutes ago      Up 31 minutes       0.0.0.0:32769->9864/tcp                          ozone-om-ha-s3_datanode_1
89f0e6a0f4c8        apache/ozone-runner:20191107-1   "/usr/local/bin/dumb…"   31 minutes ago      Up 31 minutes       0.0.0.0:9894->9872/tcp, 0.0.0.0:9884->9874/tcp   ozone-om-ha-s3_om3_1
c34627654655        apache/ozone-runner:20191107-1   "/usr/local/bin/dumb…"   31 minutes ago      Up 31 minutes       0.0.0.0:9876->9876/tcp                           ozone-om-ha-s3_scm_1
d97ff58182c1        apache/ozone-runner:20191107-1   "/usr/local/bin/dumb…"   31 minutes ago      Up 2 seconds        0.0.0.0:9892->9872/tcp, 0.0.0.0:9882->9874/tcp   ozone-om-ha-s3_om2_1


$ docker stop 81e613bda640
81e613bda640

bash-4.2$ ozone admin om getserviceroles -id=id1
proxyom1 : FOLLOWER
om2 : FOLLOWER
om3 : LEADER
```

### CODE REVIEW

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/OMProxyInfo.java [hunk 1]
```diff
@@ -21,23 +21,48 @@
...
+          rpcAddress, serviceId, nodeId);
+    }
+
+    // This issue will be a problem with docker/kubernetes world where one of
+    // the container is killed, and that OM address will be unresolved. For now
+    // skip the unresolved OM address setting it to the token service field.
+
+    // TODo: Retry rpcAddr when unresolved during getProxy which will be used
+    //  until next fail over. Fix setting of dtService also at that time.
+    try {
```
Comment (xiaoyuyao) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/OMProxyInfo.java]: should we attempt this iif rpcAddr.isUnresolved()is false?

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/OMFailoverProxyProvider.java [hunk 2]
```diff
@@ -191,20 +192,19 @@ public Text getCurrentProxyDelegationToken() {
...
-          rpcAddress.append(
-              omProxyInfoSet.getValue().getDelegationTokenService());
+      Text dtService = omProxyInfoSet.getValue().getDelegationTokenService();
 
-      if (omProxyInfos.size() != count) {
-        rpcAddress.append(",");
+      // dtService can be null when during client object creation when one of
+      // the OM configured address in unreachable.
+      if (dtService != null) {
+        rpcAddress.append(",").append(rpcAddress);
```
Comment (xiaoyuyao) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/OMFailoverProxyProvider.java]: As we discussed offline, can you add a TODO to fix the computed DTService when the dtService is temporarily avialable due to OM instance reboot, DNS issue?

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/OMProxyInfo.java (hunk 1 above)
Comment (bharatviswa504) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/OMProxyInfo.java]: Yes, good idea. done.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/OMFailoverProxyProvider.java (hunk 2 above)
Comment (bharatviswa504) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/OMFailoverProxyProvider.java]: I have added TODO in OMProxyInfo.java. Based on @arp7 offline comments removed TODO and create a Jira to resolve this.
https://issues.apache.org/jira/browse/HDDS-3137

Comment (bharatviswa504) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/OMFailoverProxyProvider.java]: I have an in-progress patch for HDDS-3233. I need to test out the scenario, and I see few in cases not resolving the address. Still looking into that part, will post the change as part of HDDS-3233.

**Pull request merged**